Source/BSGConnectivity.m

```diff
@@ -0,0 +1,115 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+#import "BSGConnectivity.h"
+
+#import <sys/socket.h>
+#import <netinet/in.h>
+#import <netinet6/in6.h>
+#import <arpa/inet.h>
+#import <ifaddrs.h>
+#import <netdb.h>
+
+@interface BSGConnectivity ()
+
+@property (nonatomic, assign) SCNetworkReachabilityRef reachabilityRef;
+@property (nonatomic, strong) dispatch_queue_t serialQueue;
+@property (nonatomic, strong) id reachability;
+
+- (void)connectivityChanged:(SCNetworkReachabilityFlags)flags;
+
+@end
+
+static void BSGConnectivityCallback(SCNetworkReachabilityRef target, SCNetworkReachabilityFlags flags, void* info) {
+    
+    BSGConnectivity *connectivity = ((__bridge BSGConnectivity *) info);
+    @autoreleasepool {
+        [connectivity connectivityChanged:flags];
+    }
+}
+
+@implementation BSGConnectivity
+
+- (instancetype)initWithURL:(NSURL *)url {
+    NSString *hostName = [url absoluteString];
+    SCNetworkReachabilityRef ref = SCNetworkReachabilityCreateWithName(NULL, [hostName UTF8String]);
+    
+    if (self = [super init]) {
+        self.reachabilityRef = ref;
+        self.serialQueue = dispatch_queue_create("com.bugsnag.cocoa", NULL);
+    }
+    return self;    
+}
+
+- (void)dealloc {
+    [self stopWatchingConnectivity];
+
+    if (self.reachabilityRef) {
+        CFRelease(self.reachabilityRef);
+        self.reachabilityRef = nil;
+    }
+
+    self.connectivityChangeBlock = nil;
+    self.serialQueue = nil;
+}
+
+/**
+ * Sets a callback with SCNetworkReachability
+ */
+- (void)startWatchingConnectivity {
+    SCNetworkReachabilityContext context = { 0, NULL, NULL, NULL, NULL };
+    context.info = (__bridge void *) self;
+
+    if (SCNetworkReachabilitySetCallback(self.reachabilityRef, BSGConnectivityCallback, &context)) {
+        if (SCNetworkReachabilitySetDispatchQueue(self.reachabilityRef, self.serialQueue)) {
+            self.reachability = self;
+        } else {
+            SCNetworkReachabilitySetCallback(self.reachabilityRef, NULL, NULL);
+        }
+    }
+}
+
+/**
+ * Stops the callback with SCNetworkReachability
+ */
+-(void)stopWatchingConnectivity {
+    SCNetworkReachabilitySetCallback(self.reachabilityRef, NULL, NULL);
+    SCNetworkReachabilitySetDispatchQueue(self.reachabilityRef, NULL);
+    self.reachability = nil;
+}
+
+- (void)connectivityChanged:(SCNetworkReachabilityFlags)flags {
+    BOOL connected = YES;
+    
+    if (!(flags & kSCNetworkReachabilityFlagsReachable)) {
+        connected = NO;
+    }
+    if (connected) { // notify via block
+        self.connectivityChangeBlock(self);
```

Comment:
Since this is optional and nil by default. Might be worth check if its nil before calling it. If `connectivityChangeBlock` is nil then this will crash.